src/Toml/Type/AnyValue.hs

```diff
@@ -30,7 +30,15 @@ instance Show AnyValue where
     show (AnyValue v) = show v
 
 instance Eq AnyValue where
-    (AnyValue val1)  == (AnyValue val2)  = isRight $ sameValue val1 val2
+    (AnyValue (Bool b1))    == (AnyValue (Bool b2))    = b1 == b2
+    (AnyValue (Integer i1)) == (AnyValue (Integer i2)) = i1 == i2
+    (AnyValue (Double f1))  == (AnyValue (Double f2))
+        | isNaN f1 && isNaN f2 && isRight (sameValue (Double f1) (Double f2)) = True
+        | otherwise = f1 == f2
```

Comment:
Okay, this piece of code requires some advanced knowledge of Haskell. I can't find any blog post about this so I will try to describe here. In `tomland` we have the following data type:

```haskell
data AnyValue = forall (t :: TValue) . AnyValue (Value t)
```

This is so-called an existential wrapper around `Value`. But basically this data type stores single field of type `Value t`. What I would like to have is `Eq` instance for `AnyValue` in the following form:

```haskell
instance Eq AnyValue where
    AnyValue val1 == AnyValue val2 = val1 == val2
```

`AnyValue` just stores `Value` and we already have `Eq` instance for `Value`, so why not reuse it, right? Unfortunately, this exact code doesn't compile. You probably tried this solution and it didn't work. The reason for this behavior is that we have `Eq` instance for `Value t` and `==` operator has the following type:

```haskell
(==) :: Value t -> Value t -> Bool
```

Which means that you can't compare `Value` parametrized by different types. And since `Value` type is GADT it actually can be parametrized by different types.

```haskell
λ: let val1 = Bool True
λ: let val2 = Double 3.5
λ: :t val1
val1 :: Value 'TBool
λ: :t val2
val2 :: Value 'TDouble
λ: val2 == val2
True
λ: val1 == val2

<interactive>:5:9: error:
    • Couldn't match type ‘'TDouble’ with ‘'TBool’
      Expected type: Value 'TBool
        Actual type: Value 'TDouble
    • In the second argument of ‘(==)’, namely ‘val2’
      In the expression: val1 == val2
      In an equation for ‘it’: it = val1 == val2
```

But, there's magic `Refl` data type (also known as _propositional equality_) in Haskell with the following definition:

```haskell
data a :~: b where
  Refl :: a :~: a
```

It's also a GADT. This data type parametrized by two type variables. But the only way to create values of this type is only when you know that types are equal. And this is where `sameValue` function comes in. `sameValue` function checks whether two `Value` of possibly different types are the same:

```haskell
λ: sameValue val2 val2
Right Refl
λ: sameValue val1 val2
Left Expected type 'Bool' but actual type: 'Double'
```

This function doesn't check that the content of `Value` is the same. We have `Eq` instance to check content of `Value`.

So, returning back to `AnyValue`: here some GHC magic involves. You can't write `val1 == val2` because the types can be different. But if you pattern-match on `Refl`, compiler will know that types of `val1` and `val2` are actually the same! And we can call `==` on this.


So this doesn't work:

```haskell
instance Eq AnyValue where
    AnyValue val1 == AnyValue val2 = val1 == val2
```

But this works:

```haskell
instance Eq AnyValue where
    AnyValue val1 == AnyValue val2 = case sameValue val1 val2 of
        Right Refl -> val1 == val2
        Left ... you can figure out that to do here by your own ;) ...
```

Reply:
Thank you for explain!
Actually I tried this one
```
(AnyValue val1)  == (AnyValue val2)  = isRight (sameValue val1 val2) && (val1 == val2)
```
And got
```
Couldn't match type ‘t1’ with ‘t’
  ‘t1’ is a rigid type variable bound by
    a pattern with constructor:
      AnyValue :: forall (t :: TValue). Value t -> AnyValue,
    in an equation for ‘==’
```
Here `Right Refl -> val1 == val2` both values are checked and `(==`) works, in my attempt  `(==)`  was parallel to checking, and therefore the error returned. That is the explain, I guess.

Reply:
Yeah, you need to perform pattern-matching manually to bring type equality into scope